From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: snek <the@snek.dev>
Date: Mon, 18 Aug 2025 09:30:28 +0200
Subject: [PATCH 5/8] Add line table options to control debuginfo output

This provides a good midpoint for embedders to have enough debug info for useful stack traces while still maintaining a small binary.
---
 config/compiler/BUILD.gn | 51 ++++++++++++++++++++++++++--------------
 1 file changed, 34 insertions(+), 17 deletions(-)

diff --git a/config/compiler/BUILD.gn b/config/compiler/BUILD.gn
index 3137854..bec888c 100644
--- a/config/compiler/BUILD.gn
+++ b/config/compiler/BUILD.gn
@@ -210,6 +210,18 @@ declare_args() {
   # away from the opt-in wexit_time_destructors config, and into the default
   # compiler config.
   is_wexit_time_destructors_default = build_with_chromium
+
+  # Limits the debuginfo that gets generated in "minimal symbols" mode to just
+  # line tables. This flag only has any effect when `symbol_level` is set to 1.
+  line_tables_only = false
+
+  # When building with `clang -g1`, a large amount of debug data is added to
+  # the final executable, mostly in the form of `DW_TAG_inlined_subroutine` tags
+  # or its COFF equivalent. Leaving out those tags reduces the size of the debug
+  # sections to a much more reasonable 5MB while still providing pretty usable
+  # stack traces when the application crashes. This flag only has any effect
+  # when `symbol_level` is set to 1.
+  no_inline_line_tables = false
 }
 
 declare_args() {
@@ -3082,10 +3094,15 @@ config("symbols") {
 config("minimal_symbols") {
   configs = []
   rustflags = []
+  cflags = []
+  ldflags = []
+
+  if (is_clang && no_inline_line_tables) {
+    cflags += [ "-gno-inline-line-tables" ]
+  }
+
   if (is_win) {
     # Functions, files, and line tables only.
-    cflags = []
-
     if (is_clang) {
       cflags += [
         # Disable putting the compiler command line into the debug info to
@@ -3095,25 +3112,17 @@ config("minimal_symbols") {
     }
     if (is_clang && use_lld && use_ghash) {
       cflags += [ "-gcodeview-ghash" ]
-      ldflags = [ "/DEBUG:GHASH" ]
+      ldflags += [ "/DEBUG:GHASH" ]
     } else {
-      ldflags = [ "/DEBUG" ]
+      ldflags += [ "/DEBUG" ]
     }
 
     # All configs using /DEBUG should include this:
     configs = [ ":win_pdbaltpath" ]
-
-    # Enable line tables for clang. MSVC doesn't have an equivalent option.
-    if (is_clang) {
-      # -gline-tables-only is the same as -g1, but clang-cl only exposes the
-      # former.
-      cflags += [ "-gline-tables-only" ]
-    }
   } else {
     if (use_debug_fission) {
       configs += [ ":split_dwarf" ]
     }
-    cflags = []
     if (is_mac && !use_dwarf5) {
       # clang defaults to DWARF2 on macOS unless mac_deployment_target is
       # at least 10.11.
@@ -3133,16 +3142,13 @@ config("minimal_symbols") {
       rustflags += [ "-Zdwarf-version=5" ]
     }
 
-    cflags += [ "-g1" ]
-
     if (is_clang && !is_tsan && !is_asan && current_os != "zos" &&
-        !use_dwarf5) {
+        !use_dwarf5 && !line_tables_only) {
       # See comment for -gdwarf-aranges in config("symbols").
       cflags += [ "-gdwarf-aranges" ]
     }
 
-    ldflags = []
-    if (is_android && is_clang) {
+    if (is_android && is_clang && !line_tables_only) {
       # Android defaults to symbol_level=1 builds, but clang, unlike gcc,
       # doesn't emit DW_AT_linkage_name in -g1 builds.
       # -fdebug-info-for-profiling enables that (and a bunch of other things we
@@ -3152,6 +3158,16 @@ config("minimal_symbols") {
       cflags += [ "-fdebug-info-for-profiling" ]
     }
 
+    if (!is_clang && line_tables_only) {
+      # On Mac/Linux using Clang/LLVM 15, the -gline-tables-only flag must be the
+      # **last** -gxxx flag. All other flags like -gno-inline-line-tables and
+      # -gno-column-info must come first, otherwise full debuginfo is produced.
+      # On Windows, `clang-cl` does not support the -g1 flag.
+      cflags += [ "-gline-tables-only" ]
+    } else {
+      cflags += [ "-g1" ]
+    }
+
     # Compressing debug info speeds up "isolate tests" step (https://crbug.com/503797275)
     # TODO(crbug.com/510291227): Enabling for linux & chromeos breaks
     # sandbox_linux_unittests's BaselinePolicy tests. Figure out why.
@@ -3161,6 +3177,7 @@ config("minimal_symbols") {
 
     asmflags = cflags
   }
+
   rustflags += [ "-Cdebuginfo=1" ]
 }
 
